Informational RFC: Alternative Planning

justincc opened the pull request:
This is an informational RFC that tries to boil down some DCP problems and put forward alternatives that are different from those in the roadmaps and planning RFC. This is done in response to being asked to write an alternative, so it's very much in the spirit of constructive engagement and with the desire of creating a better working DCP, an aim I know we all share.

Apologies if you wanted to review and you're not on the list. Please review anyway! Github limits reviewers to a max of 15.

justincc committed:
alternative planning draft

justincc committed:
fix process cartoon link

justincc committed:
fix process cartoon link for reals this time

justincc committed:
'lots more', not 'listed above'

justincc committed:
put note about reviewers

justincc committed:
amend reviewers disclaimer - github has a max of 15

justincc committed:
acknowledge that the doc criticizes rfcs but is an rfc

justincc committed:
tidy up plea paragraph

diekhans commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,308 @@
+ mandates that these require an RFC. An RFC
+takes a minimum of 3 weeks from inception to approval, though realistically RFCs of any complexity require considerably
+more time to understand the problem and align with other planned and existing technical work.
+
```
This fits with my observations.   The RFC process is very slow-moving, however, there does need to be some why to formally communicate, as one component can't change things out from another.  I don't really have any good ideas on how to improve it.

brianraymor replied:
@maniarathi and @NoopDog are currently working on an update to the RFC process, but it's not ready for community review. Part of the intention is to streamline the review process.

diekhans commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,308 @@
+the objective to implementation work stack (e.g. between developers and product owners, between product owners and 
+project leads).
+* **No overall control** - no organization is in overall control.
+* **Cultural variation** - organization's planning and development culture varies.
```
A related issue is the DCP is tightly coupled.  While the goal was loose coupling, the interfaces and behavior being loosely defined lead to tight coupling.   The shared metadata model across the components also causes tight coupling.

Commit:
adjust wording

Commit:
update process cartoon

Commit:
update process cartoon

Commit:
update process cartoon

Commit:
update process cartoon

Commit:
update process cartoon

MightyAx committed:
minor spelling changes

Changes listed incase I've changed the context:
- esepcially -> especially
- Suprisingly -> Surprisingly
- paerfect -> perfect
- adapation -> adaptation
- evangalization -> evangelization
- architectual -> architectural

sampierson commented:
This is great.  Thanks for taking the time to write this.
However long it took you to write it, it was a good investment.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+
+## Detailed Design
+
+From my understanding, the roadmaps and planning RFC describes a 70 day process where scientists, project leads, product owners, user experience
```
RE _70 day process_ - as I slacked you on August 22 following the last DCP PM meeting _The PL(s) must deliver a draft annual roadmap by August 30 which will not be based on the RFC process. Using this roadmap as a “bootstrap”, I proposed to eliminate the annual roadmap process and focus on quarterly updates/refinements to the original roadmap in time for Release planning each quarter - similar to the Technical Objectives RFC. There were no objections, so I will be rewriting that section with Cara’s review.._ I expect to complete those updates this week, now that the Planning+Execution sections of the RFC have addressed pending comments.

justincc replied:
Apologies. I was punch-drunk on getting this out. I will take that into account soon.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+more time to understand the problem and align with other planned and existing technical work.
+
+From my experience so far, if an RFC is required this means that it will take a minimum of 2 milestones (8 weeks)
+before it can be considered complete. A T-shirt estimate is only possible once the RFC is approved and there is a
```
There is nothing preventing teams from writing smaller RFC(s) and iterating towards an objective similar to AUDR1 and AUDR2. As needed, smaller scope should be negotiated to ensure that we're not spending 8 weeks on RFCs or slipping a Spike because we need one more RFC than originally planned.

justincc replied:
The point is that writing RFCs is a big overhead, there is little incentive to do smaller ones. And once you include negotiation that's another cross-DCP round of discussion which makes the process take even longer.

diekhans replied:
it would be worthwhile looking into why RFCs are such a big overhead. Is it the process or the nature of the knowledge we are trying to communicate?

Personally, I find the lack of overall context makes the process more difficult.  One either has to write up a lot of context or reviews have to have that context.  If RFCs were pull requests against as system architecture or component API documentation, it would be much lighter weight.

Architectural Decisions Records (ADR) might be a lighter-weight approach.  I don't have experience with them, so I can't say much beyond what I read on the internet.

https://niels.nu/blog/2018/lightweight-architecture-decision-records.html

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+I have grave doubts about doing down this road for two reasons:
+
+* **Prediction is hard** - predicting what software you need to create is hard. It's harder the further out you try. 
+Surprisingly soon even the roughest prediction is impossible.
```
That is a personal opinion rather than a fact. I've known many developers who can estimate and deliver on time and at quality.

justincc replied:
If that is a personal opinion then so is yours. Sure, it's easier to do such estimates in a predictable environment where all development is taking place in a single organization and site. But the point is that we have entirely the opposite here.

brianraymor replied:
I don't recall mentioning a _predictable_ environment. In fact, I was thinking about a situation where we were implementing WebSockets across Microsoft client and server while the JavaScript API (W3C) and the protocol (IETF) were in play at their respective standards organizations. Fun times.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+communication both between DCP groups with the same role (e.g. ingest, analysis, azul dev groups) and up and down
+the objective to implementation work stack (e.g. between developers and product owners, between product owners and 
+project leads).
+* **No overall control** - no organization is in overall control.
```
What role does the Oversight Committee play in this challenge?

justincc replied:
Ensuring that the aims make sense.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+
+#### Best practice is evangelized rather than mandated
+This approach recognizes that there is *cultural variation* between teams. It assumes that imposing a process will not bring
+alignment because there is *no overall control*. Teams also face different circumstances which means one size will not
```
RE *no overall control* - what is the role of the Oversight Committee in this case?

justincc replied:
A single committee also made up of people from different institutions is not project-wide control.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+ 
+#### A culture of individual empowerment
+People are afraid to take risks in this project, despite our encouragement. I've seen this many times in my relatively
+short time with the DCP. Why is this? One reason is that people find the de facto mechanism of cross-component change,
```
Can you offer some concrete examples where people were afraid to take risks?

justincc replied:
No. If people didn't want to take risks because they didn't want to deal with the process then I'm not going to name them either unless they want to be named. Either people agree from their own experience that my statement is true or they don't.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+a) Formally structured so one needs to spend a lot of time writing it.
+b) Going to be reviewed by a lot of people, you need to get it right or you'll look stupid. Both content and form.
+c) Going to generate a metric ton of discussion that will take weeks to understand and wade through.
+d) Get approved but then sit around because implementing it isn't a priority. 
```
There is an expectation that RFC(s) will be related to priorities expressed in Product or Technical Objectives which would result in their implementation.

theathorn replied:
I agree that the RFC process can be unwieldy. Unclear how to resolve an issue when one reviewer requests changes that the author doesn't agree with and others haven't commented on. How do we reach "rough consensus" to address feedback?

brianraymor replied:
It depends on the specific situation. Is the reviewer offering concrete alternatives that can be compared and contrasted? Did the author(s) address if not accommodate the issue? Do you have a specific example in mind?

In preparation for Oversight review, the Shepherd is supposed to _Summarize the review discussion for the Approvers in the top-level summary comment in the RFC pull request_ to capture such tensions.

Do you have [suggestions](https://github.com/HumanCellAtlas/dcp-community/issues) to improve the RFC process in its next version?

diekhans replied:
> In preparation for Oversight review, t

The fact that Oversight is involved in the RFC process is indicative of a problem with the RFC process.  They are far too distant from technology to be able to provide good decisions.  The advice will come in a vacuum of understanding of intricate technical details of most RFCs.

Oversite's involvement should stop at the roadmap.    If those of us who are technologists are unable to translate the roadmap into technology decisions, we have demonstrated that we are not capable of implementing the DCP.

brianraymor replied:
@diekhans - The _Oversight Committee_ is **not** involved in the RFC process. Let me refresh your memory.

When the RFC process was written, an [Oversight Review](https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0001-rfc-process.md#shepherding-rfcs) referred to the DCP PM and Technical Architecture teams:

> For software RFCs, request an Oversight review of the RFC on the HumanCellAtlas #tech-architecture slack channel. Include a link to the RFC pull request and the last call deadline
Add the RFC as an agenda item to the next DCP Architecture meeting

> For governance RFCs, request an Oversight review of the RFC on the HumanCellAtlas #dcp-project-mgmt slack channel. Include a link to the RFC pull request and the last call deadline:

When the original _Steering Committee_ reincarnated as the _Oversight Committee_, _Oversight_ became the shorthand for referring to the committee, because _OC_ was already taken by the _Organizing Committee_.  If it's confusing, then perhaps we can adopt a new name for these types of reviews.

mweiden commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+So what could be the alternatives? I am going to propose several actions which I think could be implemented separately but
+ may be most powerful if they were all to be implemented together.
+
+#### Management by aims and indicators
```
I've thought about this possibility for some time now after @tburdett first suggested it. I think there is a fundamental challenge that DCP has that other organizations who practice agile do not: _our microservices are pipelined in a deeply nested architecture_.

Compare two software architectures and their affects on coordination:

1. At a prior company, we had a "flat" architecture. The important backend services didn't regularly depend on each other and fed a back-end-for-front-end (BEFE) gateway: `ServiceA -> BEFE <- ServiceB...`. This meant that the node depth of the graph was 2 (`Service -> BEFE -> FE`) and you only ever had to change two edges in the graph to deliver new value to user. Teams could run agile scrums 90% of the time and when an edge needed to change they had a couple weeks of coordination with a predetermined migration plan.
2. Compare the prior example to the DCP where the path is `Ingest -> Upload -> DSS -> Analyisis -> Ingest -> Upload -> DSS -> {Browser, Matrix}`. The node depth here is 7. Unless the architecture changes radically, this means that there will necessarily be more coordination required to deliver value to users unless you are `{Browser, Matrix}` and already have the data you need to deliver a feature.

[A reference](https://medium.learningbyshipping.com/plans-3bcc6164e8b) to back this up:
> If you’re coordinating plans across two orgs, then timing is the only universal language — and it is a language of commitment. Two teams can only work together if they share high integrity timing — two functional teams or client/back end, etc.

Note: this is not a plug for waterfall, just an acknowledgement that, with our current architecture, we'll necessarily need to plan ahead and coordinate to deliver value to users.

justincc replied:
@mweiden I think that is a very valuable insight. However, I would draw the conclusion that having such a great node depth means that we have fundamentally the wrong architecture. No amount of process, waterfall or otherwise, will fix that. It needs some overall architecture rethinking which is near impossible because of our per-component split.

mweiden replied:
@justincc there's a bit of a catch-22 here.

Re. refactor: how can you execute a large refactor plus migration without careful coordination and planning?

Re. per-component split: the planning process is about planning, not about how we organize ourselves. See my other comment [here](https://github.com/HumanCellAtlas/dcp-community/pull/110#discussion_r320456391).

In short, options are 1) refactor the architecture (if it can be refactored) or 2) work with the current architecture. In either case, we need careful planning and coordination to move forward. Only if we've demonstrated that we have the organization and architecture to support independent development could we safely loosen our planning cycles.

mweiden commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+Nonetheless, I believe that management by aims and indicators is a management framework that it would be very valuable
+to try.
+
+#### Theme teams
```
Theme teams are not incompatible with @brianraymor's planning RFC. We could do both! 🎏

justincc replied:
Agreed.

theathorn commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+6) Theme teams grow their familiarity with the DCP as a whole. There are no longer per-component silos, though one 
+would have to avoid per-theme silos emerging.
+
+Cons:
```
Another Con: Institutions have (CZI) grants that call out component-specific work (e.g. UCSC will develop a Data Browser). Relating themes back to grant deliverables will be more challenging.

jonahcool replied:
This is a good point. Going forward, we will change the reporting to reflect progress of the DCP on the whole. It can be a single report, if each of the partners prefers this, that covers progress against the Roadmap and needs of the HCA. The goal has always been to create a single platform that meets the need of the HCA community and is flexible and well engineered. That is still very much the high level goal but recent work has clarified more specific aims and priorities. The goal of the roadmap is to help us all align to shared goals and priorities and avoid building (and reporting!) on fragmented pieces.

justincc replied:
@theathorn Yeah, that's an excellent point. Having the components embedded into the contracts considerably reduces our room for reorganization.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+local databases because it's easier than trying to get cross-DCP co-operation. We contort ourselves to match
+other component's APIs and data models because trying to get other component designs to change is just too difficult.
+These local sub-optimal choices keep piling up and introduce lots of unnecessary additional complexity to the architecture.
+
```
Another alternative might be for Technical Architecture (Technical Leads) to be responsible for the overall DCP architecture in a similar fashion as the Project Leads have recently taken on responsibility for the DCP product roadmap and strategy.  Thinking outside the box.

And there are many models for architecture reviews to steal from - [*Over the years, a number of people have asked about the details surrounding Etsy’s architecture review process.*](https://www.kitchensoap.com/2017/08/12/multiple-perspectives-on-technical-problems-and-solutions/)

justincc replied:
I think it very much remains to be seen how successful the roadmap will be. As for TAs, I think it's clear that we simply do not have the time to do this - we have truckloads of other responsibilities to actually get some kind of software written in our screwed up architecture. Equally, as I've written, there's little compelling reason other than goodwill to invest the enormous amount of energy required to change things, even if one could be sure that they would change.

TimothyTickle replied:
In my opinion, our TLs feeling they can own the architecture and be empowered to improve it is MORE important than any process or roadmap.

diekhans commented:
> @diekhans - The _Oversight Committee_ is **not** involved in the RFC process. Let me refresh your memory.

Yes! That is way confusing. Something to address in the update to the RFC process.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+
+* **Prediction is hard** - predicting what software you need to create is hard. It's harder the further out you try. 
+Surprisingly soon even the roughest prediction is impossible.
+* **Science is unpredictable** - the requirements for the HCA shift. The science changes unpredictably.
```
Can you suggest concrete examples? Is this mitigated by the Science Roadmap which informs the Product Roadmap?

justincc replied:
I haven't seen the science roadmap. You could provide a link but frankly I may not have the knowledge to make a judgement. All I know is that I see constant requirements churn and reprioritization at my level coming from scientific changes. These aren't only the science but also the organization of it. The CZI initiative, for example, which came along at short notice and is displacing Q3 work.

brianraymor replied:
Jonah presented the science roadmap at the last DCP F2F. By _CZI initiative_, do you mean the _DCP Data Submission Incentive_?

justincc replied:
Naturally I meant the DCP incentive. And I've had my own exchange with @jonahcool where I think the north stars are great but we have instituted no way of measuring whether we're meeting them and we have roadmap items that have not been linked to any north star, measurable or otherwise.

TimothyTickle replied:
The hope is the longterm road map will help smooth some of this out. A scientific roadmap is also a part of the solution, and honestly better stakeholder management on the part of the PLs. We should strive to be reactive to changes in stakeholder needs (at the consortium and user level) but having clear guidance important. This is great feedback.

brianraymor commented on rfcs/text/0000-alternative-planning.md:
```diff
@@ -0,0 +1,315 @@
+recognized and changed.
+2) It may not be possible to create metrics for some objectives. This could mean the objective is not clear enough.
+
+Nonetheless, I believe that management by aims and indicators is a management framework that it would be very valuable
```
The long-term product roadmap references "aims" from the DCP Strategy document for each Objective. If the DCP delivers the Objectives based on the Roadmap priorities, then we already have _management by aims_?

justincc replied:
Perhaps (honestly, I'm too tired to go back and check everything). But I know we have *zero* indicators in that plan. How are we going to judge that we're actually fulfilling these aims?

TimothyTickle replied:
This is a really great point, management by aims and that also we are already trending in this direction. The way the long term roadmap and associated theme/objective ownership is currently positioned agrees with the sentiment of using and tracking aims. I am glad to see this convergence.

TimothyTickle reviewed (commented):
This was a thoughtful informational RFC; I hope to offer a thoughtful (but succinct review), or at least here is the attempt ;-).

From the informational RFC, I personally focused on several takeaways:

1. The concern that the current process of defining work is waterfall in nature and needs to be able to quickly adapt to uncertainty/work refinement (a current solution for this being the agile process). From my understanding of this concern, it did not stem from the DCP not talking to users but from the amount of time the design phase requires (which is attributed to the current RFC process and negotiations required around distributed teams). I do think it is important that our process is lightweight and each step adds value to the process. There seems to be work toward refining this process and so I will leave the discussion to that space but want to acknowledge I feel the RFC process should be healthy as it is core to our work.

2. Distributed siloed teams and a suggestion for themed teams. The idea came up several times indicating our component-organized teams make it difficult to optimize engineering solutions over the DCP as a whole. Geography has the potential to slow collaboration as our overlapping window for work is minimal (1 hour). This has come up more than once and I think a PL should shepherd this to get an overall sense of possibilities. I added this to the next PM meeting. 

3. The need to adapt to unpredictable and shifting priorities. It is seemed to be expressed and, I believe is true, that the more innovative science is, the more it changes, and that we have had events that have can be viewed as interrupts. There are several efforts in place to try to better see the landscape of work ahead of us (like the long term road map and the requirement of the scientific road map in our road-mapping process). These are intended to help us always have a larger view of where we are heading but are also things that are living an can be updated. We want to be able to have a level of flexibility in working with the long term road map that allows alignment with users while balancing requests in a way that does not create chaos. I believe this is how it is coming together. I encourage the team to push back when needed in the engineering plan definition process to make sure we optimize in both directions. The PLs should also continue to improve how we protect the team from interrupts.

4. Interest in management by aims/indicators. This sounds great and aligned with recent work. Please read more in the DCP Strategy #108 PR.

5. Concern over the current architectural design and the lack of empowerment (and time) to change it. This concerns me greatly, no matter our processes, our technical leaders should lead here. This is crucial. I have also added this to the next PM meeting to identify a PL who will work with the TLs to better understand this.

justincc commented:
Hi @TimothyTickle. Thank you very much for taking the time to read the RFC and providing such a well thought out response. I very much appreciate it. To take up your points in turn.

1. Yes, from my perspective the issues aren't from lack of talking to users but the sheer turnaround time that I see in the current and proposed process. I suspect this will need something more fundamental than RFC process changes since we both don't seem to have enough cross communication (as done through RFCs) and have too many RFCs to actually review. We shall see.

2. Great that you will discuss this in PM and have a PL to look at what's possible.

3. If there is a way to protect teams more from interrupts that would be great. I suspect all or at least percentage are unavoidable, in which case a planning and work process that acknowledges and accounts for that is, in my view, critical.

4. It's great that this approach is being seriously considered and I hope it can be adopted to inform the roadmap and help us have an agile work and planning process.

5. Yes, as we know there are conversations happening around this. Hopefully we can improve in this area though, as written in this RFC, I suspect this needs some pretty fundamental changes.

barkasn commented:
@justincc Can you please set a last call for this?

justincc committed:
Update 0000-alternative-planning.md

NoopDog commented:
Obsolete
